File: GitEmailPrototype/gitemail.js
/**
 * Contains all functions that get user profile data, repositories and commits.
 */
 export class GitEmail {

    constructor (account, accountType = Account.NAME) {
        this.API_BASE = "https://api.github.com";
        this.account = account;
        this.accountType = accountType;
    }
    
    /**
     * Attempts to return a jsonified response.
     * @param {string} url_end Basically the url suffix that contains the parameters.
     * @returns The returned json from the GitHub API.
     */
    async getResponse (url_end) {
        let response = await fetch(
            this.API_BASE + "/" + url_end,
            {
                method: 'GET',
                headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json',
                }
            });
        return await response.json();
    }

    /**
     * Sets the object's account and account type.
     * @param {String} account The account name or id to set.
     * @param {Repo} accountType Defines the type of account.
     * @returns The account name.
     */
    setAccount (account, accountType) {
        this.account = account;
        this.accountType = accountType;
        return account;
    }

    /**
     * Converts the objects' account from Account.ID to Account.Name.
     * @returns The username or null.
     */
    async convertToName () {
        let response = await this.getUser();
        return (user != null) ?
            this.setAccount(response["login"], Account.NAME) :
            null;
    }

    /**
     * Fetches the user account JSON from the GitHub API.
     * @returns The user JSON.
     */
    async getUser () {
        let user = await this.getResponse(`${this.accountType}/${this.account}`);
        return (user.hasOwnProperty("message")) ? null : user;
    }

    /**
     * Fetches the users' repositories.
     * @param {Repo} repoType The type filter for accepted repositories.
     * @param {Boolean} onlyNames Converts the results strictly to their repository name.
     * @returns A list of repositories. Either each entry is JSON-type or a string.
     */
    async getUserRepos (repoType = Repo.ALL, onlyNames = false) {
        let response = await this.getResponse(`${this.accountType}/${this.account}/repos`);
        if (response.hasOwnProperty("message")){
            return null;
        }
        else {
            let repos = [];
            for (var repo in response) {
                if (repoType == Repo.ALL || repoType == response[repo]["fork"]) {
                    if (!onlyNames) repos.push(response[repo]);
                    else repos.push(response[repo]["name"]);
                }
            }
            console.log(repos);
            return repos;
        }
    }

    /**
     * Fetches a specific repositories' commits for the objects' user.
     * @param {String} repo The repository to get the commits for.
     * @returns A list of commits. Maximum amount is 30 due to public API limitations.
     */
    async getRepoCommits (repo) {
        return await this.getResponse(`repos/${this.account}/${repo}/commits`);
    } 

    /**
     * Determines the users public committing email.
     * @returns The users' public email or null.
     */
    async gitEmail () {
        let user = await this.getUser();
        if (user == null) return null;
        let repos = await this.getUserRepos(Repo.OWNED, true);
        if (repos.length == 0) {
            return true;
        }
        else {
            for (let repo in repos) {
                let commits = await this.getRepoCommits(repos[repo]);
                for (var commit in commits) {
                    if (commits[commit]["author"]["login"] == this.account) {
                        return commits[commit]["commit"]["author"]["email"];
                    }
                }
            }
            return false;
        }
    }
}


/**
 * A restricted amount of account types.
 */
export class Account {

    static ID = "user";
    static NAME = "users";

    constructor (accountType) {
        this.accountType = accountType;
    }

    toString () {
        return this.accountType;
    }

}

/**
 * A restricted amount of repository types.
 */
export class Repo {

    static ALL = null;
    static FORKED = true;
    static OWNED = false;

    constructor (repo_type) {
        this.repo_type = repo_type;
    }

    toBoolean () {
        return this.repo_type;
    }
}

File: GitEmailPrototype/gitemail.py
import requests
from typing import Optional, Union
from enum import Enum


class Account(Enum):
    ID = "user"
    NAME = "users"


class Repo(Enum):
    ALL = None
    FORKED = True
    OWNED = False


class GitEmail():
    """Retrieves a given GitHub users' email address."""

    def __init__(self, account: str, account_type: Account = Account.NAME) -> None:
        """The initial function the configures the class values.

        Args:
            account (str): The account name to get the data for.
            account_type (Account, optional): The type of the passed account string. 
            Defaults to Account.NAME.
        """

        self.api_base = "https://api.github.com"
        self.account = account
        self.account_type = account_type

    def get_response(self, url_end: str, args: dict = None) -> Optional[Union[dict, list]]:
        """Makes a get request and returns the response in a valid form.

        Args:
            url_end (str): The part of the url following the base.
            args (dict, optional): The query parameters to send with the request. Defaults to None.

        Returns:
            Optional[Union[dict, list]]: One of the valid forms that match the response JSON layout.
        """

        response = requests.get(self.api_base + url_end, params=args).json()
        try:
            response_dict = dict(response)
            if "message" in list(response_dict.keys()):
                return None
            else:
                return response_dict
        except:
            try:
                return list(response)
            except:
                return None

    def convert_to_name(self) -> Optional[str]:
        """Converts any of the valid account types to the username.

        Returns:
            Optional[str]: Returns the username if found, otherwise None.
        """

        response_dict = self.get_user()
        return response_dict["login"] if "login" in response_dict.keys() else None

    def get_user(self) -> Optional[dict]:
        """Retrieves the dictionary from GitHub representing the users account.

        Returns:
            Optional[dict]: The user account dictionary if found, else None.
        """

        return self.get_response(f"/{self.account_type.value}/{self.account}")

    def get_user_repos(self, repo_type: Repo = Repo.ALL, only_names: bool = False) -> Optional[list]:
        """Retrieves a list of the users commits to a repository. Optionally filters them by
        repository type and/ or reduces the repository dictionaries to just the names.

        Args:
            repo_type (Repo, optional): The repository type. Defaults to Repo.ALL.
            only_names (bool, optional): The option to reduce the resulting list to just the
            repository names. Defaults to False.

        Returns:
            Optional[list]: The list of the repository dictionaries connected to the users account.
        """

        repos = self.get_response(f"/{self.account_type.value}/{self.account}/repos")
        if repo_type.value != None:
            repos = [repo for repo in repos if repo["fork"] == repo_type.value]
        if only_names:
            repos = [repo["name"] for repo in repos]
        return repos

    def get_repo_commits(self, repo: str) -> Optional[list]:
        """Retrieves a list of the commits of the user to a particular repository.

        Args:
            repo (str): The repository to get the commits of.

        Returns:
            Optional[list]: The list of the commit dictionaries.
        """

        return self.get_response(f"/repos/{self.account}/{repo}/commits")

    def git_email(self) -> str:
        """Gets the GitHub Email for the given user.

        Returns:
            str: The users email.
        """

        if not self.get_user():
            print(f"User {self.account} could not be found.")
            return
        repos = self.get_user_repos(repo_type=Repo.OWNED, only_names=True)
        for repo in repos:
            for commit in self.get_repo_commits(repo):
                if commit["author"]["login"] == self.account or commit["author"]["id"] == self.account:
                    return commit["commit"]["author"]["email"]